File: tools/MewUI.VisualStudio/src/Session/W3cInput.cs
using System.Windows.Input;

namespace Aprillz.MewUI.VisualStudio.Session
{
    /// <summary>
    /// Maps WPF input identifiers to the wire convention (W3C KeyboardEvent.code strings, W3C
    /// button numbers, ctrl/shift/alt/meta modifier bits) consumed by the preview session.
    /// </summary>
    internal static class W3cInput
    {
        /// <summary>Wire modifier bits: 1 ctrl, 2 shift, 4 alt, 8 meta (WPF bit order differs).</summary>
        internal static int Modifiers(ModifierKeys modifiers) =>
            ((modifiers & ModifierKeys.Control) != 0 ? 1 : 0)
            | ((modifiers & ModifierKeys.Shift) != 0 ? 2 : 0)
            | ((modifiers & ModifierKeys.Alt) != 0 ? 4 : 0)
            | ((modifiers & ModifierKeys.Windows) != 0 ? 8 : 0);

        /// <summary>W3C button number: 0 left, 1 middle, 2 right.</summary>
        internal static int Button(MouseButton button)
        {
            switch (button)
            {
                case MouseButton.Middle: return 1;
                case MouseButton.Right: return 2;
                default: return 0;
            }
        }

        /// <summary>W3C buttons bitmask after the event: 1 left, 2 right, 4 middle.</summary>
        internal static int Buttons(MouseEventArgs args) =>
            (args.LeftButton == MouseButtonState.Pressed ? 1 : 0)
            | (args.RightButton == MouseButtonState.Pressed ? 2 : 0)
            | (args.MiddleButton == MouseButtonState.Pressed ? 4 : 0);

        /// <summary>
        /// W3C KeyboardEvent.code for a WPF key; null when unmapped. WPF reports virtual keys,
        /// which for the mapped set coincide with physical codes on standard layouts.
        /// </summary>
        internal static string KeyCode(Key key)
        {
            if (key >= Key.A && key <= Key.Z)
            {
                return "Key" + (char)('A' + (key - Key.A));
            }
            if (key >= Key.D0 && key <= Key.D9)
            {
                return "Digit" + (char)('0' + (key - Key.D0));
            }
            if (key >= Key.NumPad0 && key <= Key.NumPad9)
            {
                return "Numpad" + (char)('0' + (key - Key.NumPad0));
            }
            if (key >= Key.F1 && key <= Key.F24)
            {
                return "F" + (1 + (key - Key.F1));
            }

            switch (key)
            {
                case Key.Back: return "Backspace";
                case Key.Tab: return "Tab";
                case Key.Enter: return "Enter";
                case Key.Escape: return "Escape";
                case Key.Space: return "Space";
                case Key.Left: return "ArrowLeft";
                case Key.Up: return "ArrowUp";
                case Key.Right: return "ArrowRight";
                case Key.Down: return "ArrowDown";
                case Key.Insert: return "Insert";
                case Key.Delete: return "Delete";
                case Key.Home: return "Home";
                case Key.End: return "End";
                case Key.PageUp: return "PageUp";
                case Key.PageDown: return "PageDown";
                case Key.Add: return "NumpadAdd";
                case Key.Subtract: return "NumpadSubtract";
                case Key.Multiply: return "NumpadMultiply";
                case Key.Divide: return "NumpadDivide";
                case Key.Decimal: return "NumpadDecimal";
                default: return null;
            }
        }
    }
}

File: tools/rider-mewui/src/main/kotlin/com/aprillz/mewui/preview/W3cInput.kt
// Maps AWT input identifiers to the wire convention (W3C KeyboardEvent.code strings, W3C
// button numbers, ctrl/shift/alt/meta modifier bits) consumed by the preview session.
package com.aprillz.mewui.preview

import java.awt.event.InputEvent
import java.awt.event.KeyEvent
import java.awt.event.MouseEvent

object W3cInput {
    /** Wire modifier bits: 1 ctrl, 2 shift, 4 alt, 8 meta. */
    fun modifiers(event: InputEvent): Int =
        (if (event.modifiersEx and InputEvent.CTRL_DOWN_MASK != 0) 1 else 0) or
            (if (event.modifiersEx and InputEvent.SHIFT_DOWN_MASK != 0) 2 else 0) or
            (if (event.modifiersEx and InputEvent.ALT_DOWN_MASK != 0) 4 else 0) or
            (if (event.modifiersEx and InputEvent.META_DOWN_MASK != 0) 8 else 0)

    /** W3C button number: 0 left, 1 middle, 2 right. */
    fun button(event: MouseEvent): Int = when (event.button) {
        MouseEvent.BUTTON2 -> 1
        MouseEvent.BUTTON3 -> 2
        else -> 0
    }

    /** W3C buttons bitmask after the event: 1 left, 2 right, 4 middle. */
    fun buttons(event: MouseEvent): Int =
        (if (event.modifiersEx and InputEvent.BUTTON1_DOWN_MASK != 0) 1 else 0) or
            (if (event.modifiersEx and InputEvent.BUTTON3_DOWN_MASK != 0) 2 else 0) or
            (if (event.modifiersEx and InputEvent.BUTTON2_DOWN_MASK != 0) 4 else 0)

    /** W3C KeyboardEvent.code for an AWT key code; null when unmapped. */
    fun keyCode(keyCode: Int): String? {
        if (keyCode in KeyEvent.VK_A..KeyEvent.VK_Z) {
            return "Key" + ('A' + (keyCode - KeyEvent.VK_A))
        }
        if (keyCode in KeyEvent.VK_0..KeyEvent.VK_9) {
            return "Digit" + ('0' + (keyCode - KeyEvent.VK_0))
        }
        if (keyCode in KeyEvent.VK_NUMPAD0..KeyEvent.VK_NUMPAD9) {
            return "Numpad" + ('0' + (keyCode - KeyEvent.VK_NUMPAD0))
        }
        if (keyCode in KeyEvent.VK_F1..KeyEvent.VK_F12) {
            return "F" + (1 + (keyCode - KeyEvent.VK_F1))
        }

        return when (keyCode) {
            KeyEvent.VK_BACK_SPACE -> "Backspace"
            KeyEvent.VK_TAB -> "Tab"
            KeyEvent.VK_ENTER -> "Enter"
            KeyEvent.VK_ESCAPE -> "Escape"
            KeyEvent.VK_SPACE -> "Space"
            KeyEvent.VK_LEFT -> "ArrowLeft"
            KeyEvent.VK_UP -> "ArrowUp"
            KeyEvent.VK_RIGHT -> "ArrowRight"
            KeyEvent.VK_DOWN -> "ArrowDown"
            KeyEvent.VK_INSERT -> "Insert"
            KeyEvent.VK_DELETE -> "Delete"
            KeyEvent.VK_HOME -> "Home"
            KeyEvent.VK_END -> "End"
            KeyEvent.VK_PAGE_UP -> "PageUp"
            KeyEvent.VK_PAGE_DOWN -> "PageDown"
            KeyEvent.VK_ADD -> "NumpadAdd"
            KeyEvent.VK_SUBTRACT -> "NumpadSubtract"
            KeyEvent.VK_MULTIPLY -> "NumpadMultiply"
            KeyEvent.VK_DIVIDE -> "NumpadDivide"
            KeyEvent.VK_DECIMAL -> "NumpadDecimal"
            else -> null
        }
    }
}
